packages/contracts/contracts/ChugSplashManager.sol

```diff
@@ -241,32 +292,81 @@ contract ChugSplashManager is Owned {
-        // be and can therefore easily check if it's already populated.
-        // TODO: See if there's a better way to handle this case because it messes with the gas
-        // cost of SET_CODE/SET_STORAGE operations in a somewhat unpredictable way.
-        ChugSplashProxy proxy = getProxyByName(_action.target);
-        if (address(proxy).code.length == 0) {
-            bytes32 salt = keccak256(bytes(_action.target));
-            ChugSplashProxy created = new ChugSplashProxy{ salt: salt }(address(this));
-
-            // Could happen if insufficient gas is supplied to this transaction, should not happen
-            // otherwise. If there's a situation in which this could happen other than a standard
-            // OOG, then this would halt the entire contract.
-            // TODO: Make sure this cannot happen in any case other than OOG.
-            require(
-                address(created) != address(proxy),
-                "ChugSplashManager: ChugSplashProxy was not created correctly"
-            );
-        }
-
-        // Actually execute the action.
-        if (_action.actionType == ChugSplashActionType.SET_CODE) {
-            proxy.setCode(_action.data);
+        // Get the proxy type for the proxy that is being used for this target.
+        bytes32 proxyType = proxyTypes[proxies[_action.target]];
+
+        // Use the default proxy if this target does not have a non-standard proxy assigned to it.
+        if (proxyType == bytes32(0)) {
+            // Make sure the proxy has code in it and deploy the proxy if it doesn't. Since we're
+            // deploying via CREATE2, we can always correctly predict what the proxy address *should*
+            // be and can therefore easily check if it's already populated.
+            // TODO: See if there's a better way to handle this case because it messes with the gas
+            // cost of SET_CODE/SET_STORAGE operations in a somewhat unpredictable way.
+            Proxy proxy = getProxyByName(_action.target);
+            if (address(proxy).code.length == 0) {
+                bytes32 salt = keccak256(bytes(_action.target));
+                Proxy created = new Proxy{ salt: salt }(address(this));
+
+                // Could happen if insufficient gas is supplied to this transaction, should not happen
+                // otherwise. If there's a situation in which this could happen other than a standard
+                // OOG, then this would halt the entire contract.
+                // TODO: Make sure this cannot happen in any case other than OOG.
+                require(
+                    address(created) != address(proxy),
+                    "ChugSplashManager: Proxy was not created correctly"
+                );
+            }
+
+            // Get the address of the proxy's current implementation in case we need to set the
+            // implementation back to its original value later.
+            address implementation = proxy.implementation();
```

Comment:
Looking at this now, I think we can actually mostly merge the logic here with the logic for non-standard contracts. Have everything be owned by the `ProxyAdmin` and use the `ProxyAdmin` as the standard interface for interacting with these proxies. Slightly more expensive but it simplifies the code. For now I think it's worth trading off gas efficiency for legibility.